torch/_inductor/index_propagation.py

```diff
@@ -271,15 +292,58 @@ def inner(*args: Any, **kwargs: Any) -> IndexPropResult:
 
+    def statically_true(self, e):
+        """
+        Given some iter_ranges, return a function that given an expression, returns whether
+        it is true or false using value ranges, guard knowledge and runtime_asserts.
+
+        FIXME I think this may not be entirely right, as we may not be able to use all runtime_asserts
+              If this is an issue, just use guards in `self.axioms`.
+
+              The proper way of handling this would be to have a global shape_env that adds
+              runtime_asserts as they happen in the code. Then, it shuld be used in SimplifyIndexing
+              to perform wrap_expr and in CSEProxy.check_bounds_lazy to elide upper / lower bounds
+              also for indirect_indexing
+        """
+        evaluated = self.shape_env._maybe_evaluate_static(
+            e,
+            axioms=self.axioms,
+            var_to_range=self.var_to_range,
+        )
+        return bool(evaluated)
+
     def indirect_indexing(
         self, index: Union[Any, IndexPropVar], size: Any, check: bool = True
     ) -> Any:
-        # nb. We do index + Where(...) rather than Where(idx >= 0, idx, idx + sz) because we don't have CSE
-        #     for SymPy expressions, so we don't want to repeat idx too much
-
-        # indirect_indexing returns a sympy value, so no need to wrap in IndexPropVar here
         if isinstance(index, IndexPropVar) and index.is_symbolic:
-            # If we are turning a indirect indexing into direct, we need to wrap it.
-            index = index.value.expr
-            return index + Where(index >= 0, 0, size)
+            # If we find something we can convert into a direct indexing we do so
+            # We still need to (perhaps) wrap the expression and add bound checks
+            # We want to do this "constant folding", as we don't allow to fuse
+            # kernels into indirect indexing
+
+            expr = sympy.sympify(index.value.expr)
+
+            # TODO Perhaps move this logic to the simplify indexing pass
+            def wrap_expr(expr):
+                # Positive, negative, mixed
+                if self.statically_true(0 <= expr):
+                    return expr
+                elif self.statically_true(expr < 0):
+                    return expr + size
+                else:
+                    return Where(expr < 0, expr + size, expr)
+
+            # Sometimes it's easier to prove 0 <= expr than the weaker -size <= expr
+            can_prove_lower = self.statically_true(0 <= expr) or self.statically_true(
+                -size <= expr
+            )
+            can_prove_upper = self.statically_true(expr < size)
+            expr = wrap_expr(expr)
+            if generate_assert(check):
+                self.fallback(
+                    "check_bounds_lazy",
+                    (Boxed(expr), size),
+                    dict(lower=not can_prove_lower, upper=not can_prove_upper),
+                )
+            return Boxed(expr)
```

Comment:
Doesn't this boxing break vertical fusion as well? The same expression with and without boxing won't compare equal so won't be allowed to fuse.

Perhaps a change of tactic, why don't you just automatically put `check_bounds` in the load buffer if it doesn't use any indirect indexing? I feel like trying to pattern match the indexing expressions is fraught with difficulty.

Reply:
So we do that for all the transformed indirect indexings into direct. I think that should work, although those that are used in stores would be checked at the beginning of the kernel. But that's fine (although generates weirder code).

Then indirect indexing we could issue it via the current pattern match.

Sure, I think that should mostly work

That leaves the annoying case we have at https://github.com/pytorch/pytorch/pull/114471#discussion_r1609670881. In this case, if you don't like that change, I don't know how to do it. You have an expression that doesn't have a `tmp` and that will be generated in the loads (or lifted outside of the reduction for loop) but we need to generate the wrapping and the device_assert for it before it. I don't think this can be done without the change above, as the wrapping may depend on variables that already live in the `compute` fragment.

Reply:
IIUC we can have an indirect load where the indirect index itself is not needed for the final indexing expression. In that case I can't see why the `device_assert` needs to come before the load. If the value isn't being used then it can't lead to an IMA, so the bounds checks are merely for user feedback.

Reply:
That's fair. Will update all this on Monday.